File: tools/bench/ii1_row_major.cpp
// II-1 §5 — 같은 원소를 같은 횟수만큼 더하는데 순서만 바꾼다.
//
//   g++ -std=c++17 -O2 tools/bench/ii1_row_major.cpp -o /tmp/rm && /tmp/rm 4000
//
// 행 우선(row-major) 순회는 메모리를 앞에서 뒤로 훑는다. 캐시 라인 하나(64바이트)에
// int 16개가 실려 오므로 16번에 한 번만 메모리를 기다린다.
// 열 우선(column-major) 순회는 매번 한 행(= n*4 바이트)을 건너뛴다. 캐시 라인이
// 재사용되지 않고 TLB 도 매번 다른 페이지를 짚는다.
//
// 복잡도는 둘 다 O(n^2) 이고 명령어 수도 같다. 갈리는 것은 접근 순서뿐이다.
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <vector>
using namespace std;

int main(int argc, char** argv) {
    int n = (argc > 1) ? atoi(argv[1]) : 4000;
    // 2차원처럼 쓰지만 실제로는 1차원 연속 메모리다. a[r][c] = flat[r * n + c].
    vector<int> flat((size_t)n * n, 1);

    auto t0 = chrono::steady_clock::now();
    long long s1 = 0;
    for (int r = 0; r < n; ++r)
        for (int c = 0; c < n; ++c) s1 += flat[(size_t)r * n + c];
    auto t1 = chrono::steady_clock::now();

    long long s2 = 0;
    for (int c = 0; c < n; ++c)
        for (int r = 0; r < n; ++r) s2 += flat[(size_t)r * n + c];
    auto t2 = chrono::steady_clock::now();

    double row = chrono::duration<double>(t1 - t0).count();
    double col = chrono::duration<double>(t2 - t1).count();
    printf("n=%d (원소 %lld개, 합 %lld/%lld)\n", n, (long long)n * n, s1, s2);
    printf("행 우선  %.4f s\n", row);
    printf("열 우선  %.4f s  (%.1f배)\n", col, col / row);
    return 0;
}

File: tools/bench/ii1_row_major.py
"""II-1 §5 — Python 리스트의 리스트에서도 순회 순서가 시간을 가른다.

    python3.13 tools/bench/ii1_row_major.py 2000

C++ 쪽(`ii1_row_major.cpp`)보다 배수가 작다. Python 의 `list` 는 값이 아니라
객체 포인터를 담으므로 행 우선 순회조차 이미 포인터 한 번을 더 따라가고,
그 비용이 캐시 미스를 덮어 버리기 때문이다. 배수 자체가 이 언어 차이의 크기다.
"""

import sys
import time

n = int(sys.argv[1]) if len(sys.argv) > 1 else 2000
a = [[1] * n for _ in range(n)]

t0 = time.perf_counter()
s1 = 0
for row in a:
    for v in row:
        s1 += v
t1 = time.perf_counter()

s2 = 0
for c in range(n):
    for r in range(n):
        s2 += a[r][c]
t2 = time.perf_counter()

row_t, col_t = t1 - t0, t2 - t1
print(f"n={n} (원소 {n * n}개, 합 {s1}/{s2})")
print(f"행 우선  {row_t:.4f} s")
print(f"열 우선  {col_t:.4f} s  ({col_t / row_t:.1f}배)")
